src/data/batch_generator.h

```diff
@@ -252,18 +246,26 @@ class BatchGenerator : public RNGEngine {
 
   BatchPtr next() {
     if(bufferedBatches_.empty()) {
-      // out of data: need to get next batch from background thread
-      // We only get here if the future has been scheduled to run; it must be valid.
-      ABORT_IF(!futureBufferedBatches_.valid(), "Attempted to wait for futureBufferedBatches_ when none pending.\n"
-          "This error often occurs when Marian tries to restore the training data iterator, but the corpus has been changed or replaced.\n"
-          "If you have changed the training corpus, add --no-restore-corpus to the training command and run it again.");
-
-      bufferedBatches_ = std::move(futureBufferedBatches_.get());
-      if (bufferedBatches_.empty() // i.e., end of Epoch
-          || getSignalFlag(SIGTERM)) { // process received SIGTERM, abandon ship ...
-        return nullptr;
+      if(runAsync_) { // by default we will run in asynchronous mode
```

Comment:
Comment to self and @frankseide: this is git being bad at doing a three-way diff. The `runAsync_` part is from the `master` branch.

Reply:
@ugermann I see what you meant with the 3-way diff.